rust/src/storage/s3/mod.rs

```diff
@@ -22,10 +23,107 @@ use uuid::Uuid;
+///
+/// Available options are described below.
+///
+/// The same key shown in the table below should be used whether passing a key in the hashmap or setting it as an environment variable.
+/// Provided keys may include configuration for the S3 backend and also the optional DynamoDb lock used for atomic rename.
+///
+/// | name/key                 | description                                                                                                                                                    |
+/// | ======================== | ============================================================================================================================================================== |
+/// | AWS_REGION               | The AWS region.                                                                                                                                                |
+/// | AWS_S3_ASSUME_ROLE_ARN   | The role to assume for S3 writes.                                                                                                                              |
+/// | AWS_S3_ROLE_SESSION_NAME | The role session name to use for assume role. If not provided a random session name is generated.                                                              |
+/// | AWS_S3_LOCKING_PROVIDER  | The locking provider to use. For safe atomic rename, this should be `dynamodb`. If empty, no locking provider is used and safe atomic rename is not available. |
+///
+/// Unconsumed `extra_opts` are passed as a `HashMap` to the `dynamodb_lock` module for configuring the dynamodb client used for safe atomic rename.
+///
+/// [dynamodb_lock::DynamoDbOptions] describes the available options.
+///
+/// Two environment variables are not included as options (and not described in the table above).
+/// These must be set as environment variables when desired and are described below:
+///
+/// * AWS_ENDPOINT_URL - This variable is used specifically for testing against localstack and should be specified in the environment.
+/// * AWS_WEB_IDENTITY_TOKEN_FILE - file describing k8s configuration.
+///   env vars to satisfy https://docs.rs/rusoto_sts/0.47.0/rusoto_sts/struct.WebIdentityProvider.html#method.from_k8s_env should be provided
+pub struct S3StorageOptions {
+    region: Region,
+    assume_role_arn: Option<String>,
+    role_session_name: Option<String>,
+    use_web_identity: bool,
+    locking_provider: Option<String>,
+    extra_opts: HashMap<String, String>,
+}
+
+impl S3StorageOptions {
+    /// Creates an instance of S3StorageOptions from environment variables.
+    pub fn from_env() -> S3StorageOptions {
+        let empty_opts = HashMap::new();
+        Self::from_map(&empty_opts)
+    }
+
+    /// Creates an instance of S3StorageOptions from the given HashMap
+    pub fn from_map(options: &HashMap<String, String>) -> S3StorageOptions {
+        fn str_or_default(map: &HashMap<String, String>, key: &str, default: String) -> String {
```

Comment:
what's with the functions in your functions? As opposed to non public functions within the `impl`?

Reply:
:xzibit:

hoisted these up.